node/src/types/block.rs

```diff
@@ -603,6 +629,26 @@ impl Block {
         bincode::serialize(body)
     }
 
+    /// Check the integrity of a block by hashing its body and header
+    pub fn verify(&self) -> Result<(), BlockValidationError> {
```

Comment:
If this is supposed to be used only on the client-side, I would move it there instead of adding it as a method on the `Block` type. This method is fairly CPU-intensive (serializing, hashing twice) and should be only used in justified cases. Exposing it here, we're risking accidental programmer usage.

Reply:
Disagree, this is the right place to expose it.  We want this defined in a place that is restricted the minimum necessary set of domain types.  There is no reason that the client needs to know anything about this operation other than needing to call the `verify` method on it.  

We did a similar thing with merkle proof verification code, so there is an established pattern of doing this as well.

I think making a note in the comments would be sufficient to warn off confused programmers from using this.

Reply:
The client already knows the types and since it is the client that decides whether it's valid or not it should also have its own rules. Currently, it works as if I ask you if what you give me is correct, you say "yes" and then I trust you. That's wrong.

> We did a similar thing with merkle proof verification code, so there is an established pattern of doing this as well.

I think the pattern is wrong.

> I think making a note in the comments would be sufficient to warn off confused programmers from using this.

If there's a better way of doing it we should. Comments are not enforceable. They go out of date.

Reply:
You are could be right about this being CPU intensive, but I think that merits instrumentation.  Think we should make a JIRA ticket to look into this?  I have a couple I have to fill out today based on our action items from yesterday.

Reply:
> The client already knows the types and since it is the client that decides whether it's valid or not it should also have its own validity rules. Currently, it works as if I ask you if what you give me is correct, you say "yes" and then I trust you. That's wrong.

This is true in a sense, but this does not reflect the reality of how the library/shared code is structured or used.  One important thing to remember here is that the fields on a `Block` that the are being used for verification are not public, and this is the correct choice.  We don't want to expose these fields to downstream clients to mutate.  

Just because the validation code lives next to the type definitions in a different crate doesn't mean that the client is somehow compromised or that the trust model is broken.  You wouldn't say the same thing about all of the other places that the client is using shared code.  I think you are confusing questions of code modularity with trust.  Or alternately put, you are confusing _where_ the validation happens (in the client), versus _how_ it happens (by using a method on the type).

Reply:
More evidence that the chosen structure is better: testability.  

It is far easier to test this validation code (as well as the merkle proof validation code) with unit tests with the chosen code structure.  Putting any of this in the client means that you likely need to either wrangle around a lot of extra de/serialization machinery or write integration tests to do even basic testing.

Reply:
I don't think it would change the testing at all. The body of `fn verify(&Block, expected_hash)` would basically be the same as it is now with `self` being replaced with `&block`. Besides, my original argument was mostly about making it more difficult to use accidentally, which can be achieved by simply moving this function out from `impl Block` to a method in `client` module where it's used.

Reply:
I went ahead and instrumented this. It looks like `Block::verify` is taking ~.16 ms even when not compiling without optimization.

I wrote the following test:

```rust
    #[test]
    fn random_block_check() {
        let mut rng = TestRng::from_seed([1u8; 16]);
        let loop_iterations = 1_000;
        let mut i = 0;
        for _ in 0..loop_iterations {
            i = i + Block::random(&mut rng).height();
        }
        println!("{}", i)
    }
```

which takes **25 s 772 ms** on my machine.

Here's the same loop with `verify()` instead:

```rust
    #[test]
    fn random_block_check() {
        let mut rng = TestRng::from_seed([1u8; 16]);
        let loop_iterations = 1_000;
        for _ in 0..loop_iterations {
            Block::random(&mut rng)
                .verify()
                .expect("block hash should check");
        }
    }
```

This takes **25 s 935 ms**.  So `(25.935 - 25.772) / 1000 ~ 0.00016`

IMO, it's a micro-optimization to be concerned about people using `Block::verify` carelessly.

Reply:
Thanks for doing that @xcthulhu . Don't you think you would even get better results without `pritnln!` in the first test?

One thing that may change the results of the second benchmark though, `Block::random` will call `FinalizedBlock::random` which will create at most 10 deploy hashes. I think we hope that there will be hundreds, if not thousands of deploys in a block. Could you test with those values instead?

Reply:
> One thing that may change the results of the second benchmark though, Block::random will call FinalizedBlock::random which will create at most 10 deploy hashes. I think we hope that there will be hundreds, if not thousands of deploys in a block. Could you test with those values instead?

I'm a little busy at the moment but I would appreciate it if you did that test yeah.

Reply:
You already have all the code, you just need to change one number from 11 to 1100.

Reply:
On second thought, I suspect you are busy too (also, isn't it rather late where you are?).  We have better things to do than micro-benchmarking, since tomorrow is the last day of the sprint.

Also, I agree with @henrytill . This story is going to be jeopardized if we have to do an elaborate refactor to put this code in the client while keeping the tests.

Reply:
I run this with `#[bench]` tool since `#[test]` was not very precise – results were indistinguishable for `deploy_count=11` and `deploy_count=1000`. 

For 1000 deploy hashes the result was:
```
test types::block::tests::bench_random_block_verify               ... bench:      69,119 ns/iter (+/- 9,416)
```
and for 10 deploy hashes
```
test types::block::tests::bench_random_block_verify               ... bench:       1,802 ns/iter (+/- 194)
```

In conclusion, hash verification isn't as expensive as I expected it to be.